torch/csrc/autograd/utils/grad_layout_contract.h

```diff
@@ -13,9 +13,22 @@ namespace utils {
 inline bool obeys_layout_contract(const at::Tensor& grad, const at::Tensor& variable) {
   TORCH_INTERNAL_ASSERT(!grad.is_sparse());
   TORCH_INTERNAL_ASSERT(!variable.is_sparse());
-  return variable.is_non_overlapping_and_dense() ?
-         (grad.strides() == variable.strides()) :
-         grad.is_contiguous(at::MemoryFormat::Contiguous);
+  if (variable.is_non_overlapping_and_dense()) {
+    // Only look at stride for dimensions that are not of size 1
+    const auto& grad_sizes = grad.sizes();
+    const auto& grad_strides = grad.strides();
+    const auto& variable_strides = variable.strides();
+    for (const auto idx : c10::irange(grad_sizes.size())) {
+      if (grad_sizes[idx] != 1) {
```

Comment:
Should we also skip the check if size=0?

Reply:
Size 0 means empty Tensor. Which is a lot trickier to think about. And a lot more rare when Parameters. So I though we could ignore this case for now.